### PR TITLE
CORTX-29711:RGW:adding default parameters to cortx_rgw.conf template 

### PR DESCRIPTION
**Pull request by sachitanands**
Added default config parameters in the rgw template file.

Signed-off-by: Sachitanand Shelake <sachitanand.shelake@seagate.com>

# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Review by rahul-tripathi-git — APPROVED** [review]
Looks good to me.

**Comment by sachitanands** [comment]
Deployment testing started : 
build job : https://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/generic/job/custom-ci/5663/

**Comment by sachitanands** [comment]
Started with another testing 
docker image : https://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/generic/job/custom-ci/5669/

File: src/rgw/setup/rgw.py [hunk 1]
```diff
@@ -677,20 +676,28 @@ def _verify_backend_store_value(conf: MappedConf):
...
-        for confstore_key, config_key, default_value in config_key_mapping:
-            # fetch actual value of parameter from confstore.
-            # if config key/value is missing in confstore then use default value mentioned in const.py
-            config_value = Rgw._get_cortx_conf(conf, confstore_key, default_value)
-            Log.info(f'Setting config key :{config_key} with value:{config_value} at {client_section} section')
-            Conf.set(Rgw._conf_idx, f'{client_section}>{config_key}', str(config_value))
+        for config_key, confstore_key in config_key_mapping.items():
+            if confstore_key is None:
+                Log.info(f'config key:{config_key} not found in rgw key mapping.\
+                    hence using default value for this key')
```
Comment (cdeshmukh) [inline src/rgw/setup/rgw.py:688]: default value should be printed in such case

File: src/rgw/setup/rgw.py [hunk 2]
```diff
@@ -677,20 +676,28 @@ def _verify_backend_store_value(conf: MappedConf):
...
+            else:
+                # fetch actual value of parameter from confstore.
+                # if config key/value is missing in confstore then use default value mentioned in config file.
+                config_value = conf.get(confstore_key)
+                if config_value is None:
+                    Log.info(f'GConfig entry is missing for config key :{config_key}.\
+                        hence using default value specified in config file.')
+                    continue
+                else :
+                    Log.info(f'Setting config key :{config_key} with value:{config_value} at {client_section} section')
```
Comment (cdeshmukh) [inline src/rgw/setup/rgw.py:699]: put code with higher probability first in execution flow (code under 'else') and error handling in else case (right now under 'if' )

File: src/rgw/setup/rgw.py [hunk 3]
```diff
@@ -700,10 +707,20 @@ def _update_svc_data_path_value(conf: MappedConf, client_section: str):
         "Update svc config file with data path key which needs pre-processing values incase of default values."
         # Fetch cluster-id
         cluster_id = Rgw._get_cortx_conf(conf, const.CLUSTER_ID_KEY)
+        svc_config_file = Rgw._get_rgw_config_path(conf)
+        Rgw._load_rgw_config(Rgw._conf_idx, f'ini://{svc_config_file}')
```
Comment (cdeshmukh) [inline src/rgw/setup/rgw.py:711]: put this 'ini' config store also in some constant, so tomorrow we can plan to change it to consul and use directly

File: src/rgw/setup/rgw.py [hunk 4]
```diff
@@ -700,10 +707,20 @@ def _update_svc_data_path_value(conf: MappedConf, client_section: str):
...
         # Create data path's default value e.g. /var/lib/ceph/radosgw/<cluster-id>
         data_path_default_value = const.SVC_DATA_PATH_DEFAULT_VALUE + cluster_id
-        SVC_DATA_PATH_PARAM = [[const.SVC_DATA_PATH_CONFSTORE_KEY, const.SVC_DATA_PATH_KEY, data_path_default_value]]
+        confstore_data_path_value = conf.get(const.SVC_DATA_PATH_CONFSTORE_KEY)
+        if confstore_data_path_value is None:
+           Log.info(f'GConfig entry is missing for config key :{const.SVC_DATA_PATH_KEY}.\
+                        hence using default value specified in config file.')
+           Conf.set(Rgw._conf_idx, f'{client_section}>{const.SVC_DATA_PATH_KEY}', str(data_path_default_value))
+        else:
+           Log.info(f'Setting config key :{const.SVC_DATA_PATH_KEY} with value:{confstore_data_path_value} at {client_section} section')
```
Comment (cdeshmukh) [inline src/rgw/setup/rgw.py:722]: same comment as line 699

File: src/rgw/setup/rgw.py [hunk 5]
```diff
@@ -700,10 +707,20 @@ def _update_svc_data_path_value(conf: MappedConf, client_section: str):
...
+           Conf.set(Rgw._conf_idx, f'{client_section}>{const.SVC_DATA_PATH_KEY}', str(confstore_data_path_value))
 
-        # Updating svc config file with above data path key, value
-        Rgw._update_svc_config(conf, client_section, SVC_DATA_PATH_PARAM)
+        Conf.save(Rgw._conf_idx)
+        Log.info(f'added paramters to {client_section} successfully..')
```
Comment (cdeshmukh) [inline src/rgw/setup/rgw.py:726]: "Added config parameters to ....

Comment (sachitanands) [reply src/rgw/setup/rgw.py]: done

File: src/rgw/setup/rgw.py (hunk 4 above)
Comment (sachitanands) [reply src/rgw/setup/rgw.py]: done

File: src/rgw/setup/rgw.py (hunk 3 above)
Comment (sachitanands) [reply src/rgw/setup/rgw.py]: done

File: src/rgw/setup/rgw.py (hunk 2 above)
Comment (sachitanands) [reply src/rgw/setup/rgw.py:701]: done

File: src/rgw/setup/rgw.py (hunk 1 above)
Comment (sachitanands) [reply src/rgw/setup/rgw.py]: done

**Comment by sachitanands** [comment]
Rebuilding docker image,
Docker image : https://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/generic/job/custom-ci/5672/

**Comment by sachitanands** [comment]
Tested this changes with cortx deployment,
Docker images : 

- cortx-docker.colo.seagate.com/seagate/cortx-all:2.0.0-5672-custom-ci
- cortx-docker.colo.seagate.com/seagate/cortx-rgw:2.0.0-5672-custom-ci
- cortx-docker.colo.seagate.com/seagate/cortx-data:2.0.0-5672-custom-ci

cluster deploy job : https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/RGW/job/setup-cortx-rgw-cluster/2164/

**Comment by sachitanands** [comment]
Below are the default parameters added in the template file. 
image

File: conf/cortx_rgw.conf [hunk 6]
```diff
@@ -9,6 +9,21 @@
     motr tracing enabled = true
     rgw enable apis = s3, s3website, swift, swift_auth, admin, sts, iam, notifications
     rgw cache enabled = false
+    rgw thread pool size = 10
+    rgw max concurrent request = 20
```
Comment (sachitanands) [inline conf/cortx_rgw.conf:13]: @andriytk , @sachinpunadikar  please suggest default value for 'rgw max concurrent request' parameter

Comment (sachinpunadikar) [reply conf/cortx_rgw.conf:13]: As per Andriy, below value should be 
rgw max concurrent request = 1024 (the default mentioned in RGW doc)

**Review by sachinpunadikar — CHANGES_REQUESTED** [review]
work on comments

File: conf/cortx_rgw.conf (hunk 6 above)
Comment (cdeshmukh) [reply conf/cortx_rgw.conf:13]: Thanks for confirming, @sachitanands lets go ahead with 1024 default value

Comment (sachitanands) [reply conf/cortx_rgw.conf]: thanks Sachin Chetan .. i have updated this 1024 as default value.

**Comment by sachitanands** [comment]
> work on comments

done.

**Comment by sachitanands** [comment]
Started final testing with default value as 1024 for rgw_max_concurrent_request parameter
docker image job : https://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/generic/job/custom-ci/5698/

**Pull request merged**